File: day18/day18.hs
module Main where
import Data.List ( foldl' )
import Data.List.Split ( splitOn )
import Data.HashSet ( HashSet, member, fromList, insert, union, toList )

parseLine :: String -> (Int, Int, Int)
parseLine line = (x, y, z)
  where [x, y, z] = map (read :: String -> Int) (splitOn "," line)

neighbors :: (Int, Int, Int) -> [(Int, Int, Int)]
neighbors (x, y, z) = [(x+1, y, z), (x-1, y, z), (x, y+1, z), (x, y-1, z), (x, y, z+1), (x, y, z-1)]

freeNeighbors :: (Int, Int, Int) -> [(Int, Int)] -> HashSet (Int, Int, Int) -> [(Int, Int, Int)]
freeNeighbors coord bounds droplet = [(x', y', z') | (x', y', z') <- neighbors coord,
                                                                    x' >= fst (head bounds) && x'<= snd (head bounds),
                                                                    y' >= fst (bounds !! 1) && y' <= snd (bounds !! 1),
                                                                    z' >= fst (bounds !! 2) && z' <= snd (bounds !! 2),
                                                                    not (coord `member` droplet)]

surfaceArea :: [(Int, Int, Int)] -> HashSet (Int, Int, Int) -> Int -> Int
surfaceArea [] _ total = total
surfaceArea (x:xs) seen total = surfaceArea xs (insert x seen) z
    where z = total + 6 - 2 * length (filter (`member` seen) (neighbors x))

findExterior :: [(Int, Int, Int)] -> HashSet (Int, Int, Int) -> HashSet (Int, Int, Int) -> [(Int, Int)] -> HashSet (Int, Int, Int)
findExterior [] _ v _ = v
findExterior queue droplet visited bounds = findExterior (tail queue ++ z) droplet (visited `union` fromList z) bounds
    where z = filter (\x -> not (x `member` visited)) (freeNeighbors (head queue) bounds droplet)

main :: IO ()
main = do
    rawData <- readFile "input.txt"
    let coords = map parseLine (filter (/= "") (splitOn "\n" rawData))
        coordSet = fromList coords
        sArea = surfaceArea coords (fromList []) 0
        bounds = [(-1, maximum [x+1 | (x, _, _) <- coords]),
                  (-1, maximum [y+1 | (_, y, _) <- coords]),
                  (-1, maximum [z+1 | (_, _, z) <- coords])]
        exterior = findExterior [(-1, -1, -1)] coordSet (fromList [(-1, -1, -1)]) bounds
        interior = [(x, y, z) | x <- [0..snd (head bounds)],
                                y <- [0..snd (bounds !! 1)],
                                z <- [0..snd (last bounds)],
                                not $ (x, y, z) `member` exterior,
                                not $ (x, y, z) `member` coordSet]
    putStrLn $ "Part 1:" ++ show sArea
    putStrLn $ "Part 2:" ++ show (sArea - surfaceArea interior (fromList []) 0 )

File: day18/day18.rs
use std::collections::{HashSet, VecDeque};

fn neighbors(cube: &Vec<i64>) -> Vec<Vec<i64>> {
    return vec![
        vec![cube[0] + 1, cube[1], cube[2]],
        vec![cube[0], cube[1] + 1, cube[2]],
        vec![cube[0], cube[1], cube[2] + 1],
        vec![cube[0] - 1, cube[1], cube[2]],
        vec![cube[0], cube[1] - 1, cube[2]],
        vec![cube[0], cube[1], cube[2] - 1],
    ];
}

fn surface_area(coords: &Vec<Vec<i64>>) -> i64 {
    let mut seen: HashSet<Vec<i64>> = HashSet::new();
    let mut area = 0;
    for cube in coords.iter() {
        area += 6;
        seen.insert(cube.to_vec());
        for neighbor in neighbors(cube) {
            if seen.contains(&neighbor) {
                area -= 2;
            }
        }
    }
    area
}

fn free_neighbors(
    pos: &Vec<i64>,
    droplet: &HashSet<Vec<i64>>,
    min: &Vec<i64>,
    max: &Vec<i64>,
) -> Vec<Vec<i64>> {
    let mut free_neighbors: Vec<Vec<i64>> = Vec::new();
    for neighbor in neighbors(pos).iter() {
        if !droplet.contains(neighbor)
            && neighbor
                .iter()
                .enumerate()
                .all(|(i, &x)| x <= max[i] && x >= min[i])
        {
            free_neighbors.push(neighbor.to_vec());
        }
    }
    free_neighbors
}

fn find_interior(coords: &Vec<Vec<i64>>) -> Vec<Vec<i64>> {
    let droplet: HashSet<Vec<i64>> = HashSet::from_iter(coords.iter().cloned());
    let min = vec![-1, -1, -1];
    let max = vec![
        coords.iter().map(|x| x[0]).max().unwrap() + 1,
        coords.iter().map(|x| x[1]).max().unwrap() + 1,
        coords.iter().map(|x| x[2]).max().unwrap() + 1,
    ];

    let mut queue: VecDeque<Vec<i64>> = VecDeque::from(vec![min.to_vec()]);
    let mut visited: HashSet<Vec<i64>> = HashSet::new();
    visited.insert(min.to_vec());
    while queue.len() > 0 {
        let v = queue.pop_front().unwrap();
        for neighbor in free_neighbors(&v, &droplet, &min, &max) {
            if !visited.contains(&neighbor) {
                visited.insert(neighbor.to_vec());
                queue.push_back(neighbor);
            }
        }
    }

    let mut interior: Vec<Vec<i64>> = Vec::new();
    for i in 0..max[0] {
        for j in 0..max[1] {
            for k in 0..max[2] {
                if !visited.contains(&vec![i, j, k]) && !droplet.contains(&vec![i, j, k]) {
                    interior.push(vec![i, j, k]);
                }
            }
        }
    }

    interior
}

fn main() {
    let raw_data = std::fs::read_to_string("input.txt").unwrap();
    let coords: Vec<Vec<i64>> = raw_data
        .trim()
        .split("\n")
        .map(|x| x.split(",").map(|y| y.parse().unwrap()).collect())
        .collect();
    let total_area = surface_area(&coords);
    println!("part 1 {:?}", total_area);
    let interior = find_interior(&coords);
    let interior_area = surface_area(&interior);
    println!("part 2 {:?}", total_area - interior_area);
}
